parser/ast/procedure_test.go

```diff
@@ -0,0 +1,149 @@
+			"select a.id,a.username,a.password,a.age,a.sex,ad.score from user a right join user_score ad on a.id = ad.user_id where a.id > 10 and a.id < 50;" +
+			"select a.id,a.username,a.password,a.age,a.sex,ad.score from user a left join user_score ad on a.id = ad.user_id where a.id in (select user_id from user_score where score > 90 and score < 99 ) " +
+			"union select a.id,a.username,a.password,a.age,a.sex,ad.score from user a left join user_score ad on a.id = ad.user_id where a.id in (select user_id from user_score where score > 30 and score < 70 ); end;",
+		`create procedure proc_2() begin select * from t1; if i > 1 then select 2 ; end if ;end`,
+		`create procedure proc_2() begin select * from t1; if i > 1 then select 2; else select id from t2  ; end if ; end`,
+		`create procedure proc_2() begin select * from t1; if i > 1 then select 2; elseif i = 3 then select 4; else select 5 ; end if; end`,
+		`create procedure proc_2(id int) begin while id < 10 do set id = id + 1; select 1; end while; end`,
+		`create procedure proc_2() begin declare test1 CURSOR for select 1; end;`,
+		`create procedure proc_2() begin declare test1 CURSOR for select 1; select 1;open test1; end;`,
+		`create procedure proc_2() begin declare a int;declare test1 CURSOR for select 1; select 1;open test1; fetch test1 into a; close test1;end;`,
+		`create procedure proc_2() begin declare a int;declare exit handler for 1111 select 1 ; end;`,
+		`create procedure proc_2() begin declare a int;declare exit handler for 1111,1112 select 1 ; end;`,
+		`create procedure proc_2() begin declare a int;declare exit handler for SQLWARNING,NOT FOUND,SQLEXCEPTION select 1 ; end;`,
+		`create procedure proc_2() begin declare a int;declare continue handler for SQLWARNING,NOT FOUND,SQLEXCEPTION select 1 ; end;`,
+	}
+	for _, testcase := range testcases {
+		stmt, _, err := p.Parse(testcase, "", "")
+		require.NoError(t, err)
+		_, ok := stmt[0].(*ast.ProcedureInfo)
+		require.True(t, ok)
+	}
+}
+
+func TestShowCreateProcedure(t *testing.T) {
+	p := parser.New()
+	stmt, _, err := p.Parse("show create procedure proc_2", "", "")
+	require.NoError(t, err)
+	_, ok := stmt[0].(*ast.ShowStmt)
+	require.True(t, ok)
+	stmt, _, err = p.Parse("drop procedure proc_2", "", "")
+	require.NoError(t, err)
+	_, ok = stmt[0].(*ast.DropProcedureStmt)
+	require.True(t, ok)
+}
+
+func TestProcedureVisitor(t *testing.T) {
+	sqls := []string{
+		"create procedure proc_2(in id bigint,in id2 varchar(100),in id3 decimal(30,2)) begin declare s varchar(100) DEFAULT FROM_UNIXTIME(1447430881);select s;SELECT * FROM `t1`;SELECT * FROM `t2`;INSERT INTO `t1` VALUES (111);END;",
+		"show create procedure proc_2;",
+		"drop procedure proc_2;",
+	}
+	parse := parser.New()
+	for _, sql := range sqls {
+		stmts, _, err := parse.Parse(sql, "", "")
+		require.NoError(t, err)
+		for _, stmt := range stmts {
+			stmt.Accept(visitor{})
+			stmt.Accept(visitor1{})
+		}
+	}
+}
+
+func TestProcedureRestore(t *testing.T) {
```

Comment:
Maybe we should add more testcases to cover the new syntax added by you, such as `out`, `input`, `contiue` ....

Reply:
done